File: chick/src/main/java/capsule/trunk/swipe/SwipeCallback.java
package capsule.trunk.swipe;

import android.graphics.Canvas;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.support.v7.widget.helper.ItemTouchHelper;

/**
 * 描 述：
 * 作 者：Vegeta Yu
 * 时 间：2017/9/14 18:46
 */
public class SwipeCallback extends ItemTouchHelper.Callback {

  @Override
  public int getMovementFlags(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder) {
    int dragFlags;
    int swipeFlags;
    if (isLinearVerticalLayout(recyclerView)) {
      dragFlags = 0;
      swipeFlags = ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT;
    } else {
      throw new RuntimeException(
          "Swipe gesture must be act on linear vertical RecyclerView!Set LinearLayoutManager to your RecyclerView!");
    }
    return makeMovementFlags(dragFlags, swipeFlags);
  }

  @Override public boolean onMove(RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder,
      RecyclerView.ViewHolder target) {
    return false;
  }

  @Override public void onSwiped(RecyclerView.ViewHolder viewHolder, int direction) {

    //List list = ((Adapter) recyclerView.getAdapter()).getData();

  }

  @Override
  public void onChildDraw(Canvas c, RecyclerView recyclerView, RecyclerView.ViewHolder viewHolder,
      float dX, float dY, int actionState, boolean isCurrentlyActive) {
    //if (actionState == ItemTouchHelper.ACTION_STATE_SWIPE) {
    //  if (Math.abs(dX) <= recyclerView.getWidth()) {
    //    viewHolder.itemView.scrollTo(-(int) dX, 0);
    //  }
    //}
    super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive);

  }

  private boolean isLinearVerticalLayout(RecyclerView recyclerView) {
    //init LayoutManager type
    RecyclerView.LayoutManager layoutManager = recyclerView.getLayoutManager();
    if (layoutManager instanceof LinearLayoutManager) {
      if (layoutManager instanceof GridLayoutManager) {
        //GridLayoutManager
        //layout_type = LAYOUT_GRID;
        return false;
      } else {
        if (((LinearLayoutManager) layoutManager).getOrientation()
            == LinearLayoutManager.VERTICAL) {
          //LinearLayoutManager -- vertical
          //layout_type = LAYOUT_VERTICAL;
          return true;
        } else {
          //LinearLayoutManager -- horizontal
          //layout_type = LAYOUT_HORIZONTAL;
          return false;
        }
      }
    } else {
      if (((StaggeredGridLayoutManager) layoutManager).getOrientation()
          == StaggeredGridLayoutManager.VERTICAL) {
        //StaggeredGridLayoutManager -- vertical
        //layout_type = LAYOUT_STAGGERED_GRID_VERTICAL;
        return false;
      } else {
        //StaggeredGridLayoutManager -- horizontal
        //layout_type = LAYOUT_STAGGERED_GRID_HORIZONTAL;
        return false;
      }
    }
  }
}

File: frogx/src/main/java/yuluyao/frog/swipe/SwipeCallback.kt
package yuluyao.frog.swipe

import android.graphics.Canvas

import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager

/**
 * 描 述：
 * 作 者：Vegeta Yu
 * 时 间：2017/9/14 18:46
 */
class SwipeCallback : ItemTouchHelper.Callback() {

  override fun getMovementFlags(recyclerView: RecyclerView,
                                viewHolder: RecyclerView.ViewHolder): Int {
    val dragFlags: Int
    val swipeFlags: Int
    if (isLinearVerticalLayout(recyclerView)) {
      dragFlags = 0
      swipeFlags = ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT
    } else {
      throw RuntimeException(
        "Swipe gesture must be act on linear vertical RecyclerView!Set LinearLayoutManager to your RecyclerView!")
    }
    return makeMovementFlags(dragFlags, swipeFlags)
  }

  override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder,
                      target: RecyclerView.ViewHolder): Boolean {
    return false
  }

  override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {

    //List list = ((ChickAdapter) recyclerView.getAdapter()).getData();

  }

  override fun onChildDraw(c: Canvas, recyclerView: RecyclerView,
                           viewHolder: RecyclerView.ViewHolder,
                           dX: Float, dY: Float, actionState: Int, isCurrentlyActive: Boolean) {
    //if (actionState == ItemTouchHelper.ACTION_STATE_SWIPE) {
    //  if (Math.abs(dX) <= recyclerView.getWidth()) {
    //    viewHolder.itemView.scrollTo(-(int) dX, 0);
    //  }
    //}
    super.onChildDraw(c, recyclerView, viewHolder, dX, dY, actionState, isCurrentlyActive)

  }

  private fun isLinearVerticalLayout(recyclerView: RecyclerView): Boolean {
    //init LayoutManager type
    val layoutManager = recyclerView.layoutManager
    return if (layoutManager is LinearLayoutManager) {
      if (layoutManager is GridLayoutManager) {
        //GridLayoutManager
        //layout_type = LAYOUT_GRID;
        false
      } else {
        if (layoutManager.orientation == LinearLayoutManager.VERTICAL) {
          //LinearLayoutManager -- vertical
          //layout_type = LAYOUT_VERTICAL;
          true
        } else {
          //LinearLayoutManager -- horizontal
          //layout_type = LAYOUT_HORIZONTAL;
          false
        }
      }
    } else {
      if ((layoutManager as StaggeredGridLayoutManager).orientation == StaggeredGridLayoutManager.VERTICAL) {
        //StaggeredGridLayoutManager -- vertical
        //layout_type = LAYOUT_STAGGERED_GRID_VERTICAL;
        false
      } else {
        //StaggeredGridLayoutManager -- horizontal
        //layout_type = LAYOUT_STAGGERED_GRID_HORIZONTAL;
        false
      }
    }
  }
}
